keras/engine/training.py

```diff
@@ -3258,6 +3258,24 @@ def concat(tensors, axis=0):
   return tf.concat(tensors, axis=axis)
 
 
+def potentially_variable_concat(tensors):
+  """Concats `tensor`s along `axis`. When non-batch dimensions are variable, a RaggedTensor is returned instead."""
+  if isinstance(tensors[0], tf.SparseTensor):
+    return tf.sparse.concat(axis=0, sp_inputs=tensors)
+  non_batch_shapes = tf.stack([tf.shape(tensor)[1:] for tensor in tensors])
+  constant_dims = tf.math.reduce_all(non_batch_shapes == non_batch_shapes[:1], axis=0)
+  if tf.math.reduce_all(constant_dims).numpy().item():  # All non-batch dimensions are constant
+    return tf.concat(tensors, axis=0)
+  # First, identify constant inner dimensions by finding the rightmost dimension that is not constant
+  constant_inner_dimensions = constant_dims.numpy().tolist()[::-1].index(False)
+  # If there are constant inner dimensions, define a constant inner shape
+  if constant_inner_dimensions == 0:
+    constant_inner_shape = None
+  else:
+    constant_inner_shape = tensors[0].shape[-constant_inner_dimensions:]
+  return tf.ragged.constant([tensor.numpy() for tensor in tensors], inner_shape=constant_inner_shape).merge_dims(0, 1)
```

Comment:
I added the switch to `numpy()` because I struggled to make this work any other way - `tf.ragged.constant` does not accept `tf.Tensor` inputs gracefully, and other methods that do like `tf.ragged.stack` often threw errors when the batch dimensions varied.

Reply:
> tf.ragged.constant does not accept tf.Tensor inputs gracefully

It's really strange that it works with numpy arrays but not tf.Tensors. That's an inconsistency in the ragged API that we should address...

Reply:
I think the cause is that `tf.ragged.constant` is not really intended for this use-case - there are other methods like `tf.RaggedTensor.from_tensor()` that are. However, I couldn't figure out any approach using those methods that preserved constant inner dimensions and didn't crash in edge cases.